lib/lib-dynamodb/README.md

```diff
@@ -0,0 +1,163 @@
+| B               | Binary     | Uint8Array, Buffer, File, Blob...      |
+| M               | Map        | { [key: string]: \<ATTRIBUTE_VALUE\> } |
+| L               | List       | \<ATTRIBUTE_VALUE\>[]                  |
+| N               | Number     | number, BigInt                         |
+
+The DynamoDB document client should abstract the attribute values as follows in
+both input and output:
+
+```json
+{ "KEY_NAME": VALUE }
+```
+
+## Usage
+
+To create document client you need to create DynamoDB client first as follows:
+
+```js
+import { DynamoDBClient } from "@aws-sdk/client-dynamodb"; // ES6 import
+// const { DynamoDBClient } = require("@aws-sdk/client-dynamodb"); // CommonJS import
+
+// Bare-bones DynamoDB Client
+const client = new DynamoDBClient({});
+```
+
+```js
+import { DynamoDB } from "@aws-sdk/client-dynamodb"; // ES6 import
+// const { DynamoDB } = require("@aws-sdk/client-dynamodb"); // CommonJS import
+
+// Full DynamoDB Client
+const client = new DynamoDB({});
+```
+
+The bare-bones clients are more modular. They reduce bundle size and improve
+loading performance over full clients as explained in blog post on
+[modular packages in AWS SDK for JavaScript](https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/).
+
+### Constructor
+
+Once DynamoDB client is created, you can either create the bare-bones
+document client or full document client as follows:
+
+```js
+import { DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb"; // ES6 import
+// const { DynamoDBDocumentClient } = require("@aws-sdk/lib-dynamodb"); // CommonJS import
+
+// Bare-bones document client
+const ddbDocClient = DynamoDBDocumentClient.from(client); // client is DynamoDB client
+```
+
+```js
+import { DynamoDBDocument } from "@aws-sdk/lib-dynamodb"; // ES6 import
+// const { DynamoDBDocument } = require("@aws-sdk/lib-dynamodb"); // CommonJS import
+
+// Full document client
+const ddbDocClient = DynamoDBDocument.from(client); // client is DynamoDB client
+```
+
+### Configuration
+
+The configuration for marshalling and unmarshalling can be sent as an optional
+second parameter during creation of document client as follows:
+
+```js
+const marshallOptions = {
+  // Whether to automatically convert empty strings, blobs, and sets to `null`.
+  convertEmptyValues: false; // false, by default.
+  // Whether to remove undefined values while marshalling.
+  removeUndefinedValues: false; // false, by default.
+  // Whether to convert typeof object to map attribute.
+  convertClassInstanceToMap: false; // false, by default.
+};
+
+const unmarshallOptions = {
+  // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
+  wrapNumbers: false; // false, by default.
+};
+
+const translateConfig = { marshallOptions, unmarshallOptions };
+
+const client = new DynamoDBClient({});
+const ddbDocClient = DynamoDBDocument.from(client, translateConfig);
+```
+
+### Calling operations
```

Comment:
It would be helpful if add an example of destroying client. We should underline that users are suppose to call `destroy()` on the basic client not the doc client.

Reply:
Adde in [`855779c` (#2097)](https://github.com/aws/aws-sdk-js-v3/pull/2097/commits/855779c87d327242c55d20a79d5c12997785743d)

Should `destroy()` on document client throw an error instead of being a no-op?

Reply:
Why not have the document client forward the destroy to the client?

Reply:
@kyeotic Technically the basic client is decoupled with the document client. They can be used at the same time(tho this would rare). So destroying one that would destroy another would be error-prone.

@trivikr I think it makes sense to throw an error here. It would be even nicer to make `destroy()` accept a flag that confirms relaying the destroy to basic client. But I'm not sure whether it's doable with overriding.

Reply:
I think destroying the other makes sense for most cases. I think the standard case will be that if you make a Document Client then that's what you plan to use; you likely only even made the DB Client in order to pass it to the Document Client. You might not even have saved a reference to the original DB client.

I dont see how this is any different than the normal situation where a single DB client is constructed for an entire app. If one part of the app destroys the DB client then other parts of the app, possibly unaware of this, will hit an error trying to use it. Destruction needs to be coordinated whether you are using a Document Client or not.

Reply:
- A) No-op allows any generic cleanup abstraction to uselessly call `destroy()` with no adverse affects, but may mislead someone into believing that they're destroying the underlying client at the same time.

- B) Throwing an error prevents someone from calling `destroy()` for no reason due to an incomplete understanding of how to use the client.

I say A is preferred over B. Rationale being that as a library user I had to have separately _created_ the underlying client, so I am unlikely to expect that some other object I'm passing it to is going to destroy it, so the risk of calling `destroy()` and expecting the basic client to be destroyed is low. In addition, if I'm relying on some generic automation to call `destroy()` on objects like this, for one of them to suddenly start throwing means I have to now work around that exceptional behaviour. Finally, it's not like calling a no-op function is burning any cycles so there's no real downside other than potentially misleading inattentive users.

Reply:
C) Destroy the underlying client; do not throw an error

I agree A is better than B. Both are inferior to C, as C reflects the intent of the coder. If I call `destory` I expect destruction to occur. This isn't a problem of attention; doing anything besides destroying the client violates the [Principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

If I have to keep the underlying client around to destroy it, even if I only plan on using the Document Client, then I need to keep an extra object reference. If the construction site is different than the desctruction site then I also need to _pass around_ two references: one that I want to use, and one that I only need for `destroy`. Doesn't that seem a bit silly?

Reply:
@kyeotic I disagree with you on the Principle– I think having a secondary object destroying its input would be more of a shock. But I do agree it seems silly to hang onto a ref for the basic client for no reason other than cleanup.

Reply:
Ok, I can see how destroying the input might surprise inattentive users (😛 ). If that's the case though, `destroy` on the Document Client _at least_ needs to release the reference its holding to the DB client.

I still think it should destroy the underlying client though, as I expect 99% of users who make a Document Client to not use the DB Client.

Reply:
I'll implement the option

> D) The `destroy()` call on Document Client releases the reference the client.

This way:
* The Document Client is destroyed, as it can no longer make the calls.
* The DynamoDB Client is still available if user wants to use elsewhere.

Reply:
I came across a bug where calling `.destroy()` doesn't actually destroy the clients which is reported at https://github.com/aws/aws-sdk-js-v3/issues/2105

I'll leave the `destroy()` as a no-op as of now for document client, and create a follow-up request for the remove DynamoDB client reference which would require additional updates to internal code base.